AcademyChallenge/Presentation/Scenes/EmojiListViewController.swift

```diff
@@ -6,202 +6,84 @@
 //
 
 import UIKit
+//
+import RxSwift
 
+public protocol EmojiListViewControlerDelegate: AnyObject {
+    func navigateToFirstPage()
+}
 
-class EmojiListViewController: UIViewController {
-	
-	var emoji: [Emoji]?
-	
-	//var strong = MockedDataSource()
-	
-	weak var emojiService: EmojiService?
-	
-	//var emojiServer: LiveEmojiStorage?
-	
-	lazy var collectionView: UICollectionView = {
-			let v = UICollectionView(frame: .zero, collectionViewLayout: UICollectionViewFlowLayout())
-			return v
-		}()
-	
-	
-		
-	
+class EmojiListViewController: BaseGenericViewController<EmojiView> {
+    public weak var delegate: EmojiListViewControlerDelegate?
+    var emoji: [Emoji]?
+    var viewModel: EmojiListViewModel?
     override func viewDidLoad() {
         super.viewDidLoad()
-		
-        setUpViews()
-		addViewToSuperview()
-		setUpConstrains()
-		
-	}
-	private func setUpViews() {
-		
-		setUpCollecionView()
-	}
-	
-
-	private func addViewToSuperview(){
-		
-		view.addSubview(collectionView)
-	}
-	
-	private func setUpConstrains() {
-		
-		collectionView.translatesAutoresizingMaskIntoConstraints = false
-		
-		NSLayoutConstraint.activate([
-			collectionView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
-			collectionView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
-			collectionView.topAnchor.constraint(equalTo: view.topAnchor),
-			collectionView.bottomAnchor.constraint(equalTo: view.bottomAnchor),
-		
-		
-		])
-	}
-	
-	private func setUpCollecionView() {
-		
-	
-		
-		let layout = UICollectionViewFlowLayout()
-		layout.scrollDirection = .vertical
-		
-		layout.minimumLineSpacing = 8
-		layout.minimumInteritemSpacing = 4
-		
-		collectionView = .init(frame: .zero, collectionViewLayout: layout)
-		
-		collectionView.backgroundColor = UIColor.appColor(.primary)
-		
-		collectionView.register(CollectionViewCell.self, forCellWithReuseIdentifier: "cell")
-		
-		collectionView.delegate = self
-		collectionView.dataSource = self
-	}
-	
-	
-	override func viewWillAppear(_ animated: Bool) {
-		
-		super.viewDidAppear(animated)
-		
-		emojiService?.fetchEmojis({ [weak self] (result: Result<[Emoji],Error>) in
-			switch result{
-			case .success(let success):
-				self?.emoji = success
-				self?.emoji?.sort()
-				DispatchQueue.main.async { [weak self] in
-					self?.collectionView.reloadData()
-				}
-				
-			case .failure(let failure):
-				print("Failure: \(failure)")
-			}
-			
-
-		})
-		
-		
-	}
-	
-
+        genericView.collectionView.dataSource = self
+    }
+    override func viewWillAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        viewModel?.getEmojisList()
+            .observe(on: MainScheduler.instance)
+            .subscribe(onSuccess: { emojis in
```

Comment:
this `subscribe` may lead to some retain cycles, a `[weak self]` could help to avoid that